src/rdkafka.c

```diff
@@ -270,7 +279,463 @@ void rd_kafka_log0 (const rd_kafka_conf_t *conf,
+                        rd_snprintf(errstr, errstr_size,
+                                "SASL/OAUTHBEARER extension keys must only consist of A-Z or a-z characters: %s (%c)",
+                                key, *c);
+                        return -1;
+                }
+                ++c;
+        }
+        return 0;
+}
+
+static int check_oauthbearer_extension_value(const char *value,
+                char *errstr, size_t errstr_size) {
+        /*
+         * https://tools.ietf.org/html/rfc7628#section-3.1
+         * value          = *(VCHAR / SP / HTAB / CR / LF )
+         * 
+         * https://tools.ietf.org/html/rfc5234#appendix-B.1
+         * VCHAR          =  %x21-7E  ; visible (printing) characters
+         * SP             =  %x20     ; space
+         * HTAB           =  %x09     ; horizontal tab
+         * CR             =  %x0D     ; carriage return
+         * LF             =  %x0A     ; linefeed
+         */
+        const char *c = value;
+        while (*c != '\0') {
+                if (!(*c >= '\x21' && *c <= '\x7E') && *c != '\x20' && *c != '\x09' && *c != '\x0D' && *c != '\x0A') {
+                        rd_snprintf(errstr, errstr_size,
+                                "SASL/OAUTHBEARER extension values must only consist of space, horizontal tab, CR, LF, and visible characters (%%x21-7E): %s (%c)",
+                                value, *c);
+                        return -1;
+                }
+                ++c;
+        }
+        return 0;
+}
+
+int rd_kafka_oauthbearer_set_token(rd_kafka_t *rk,
+                const char *token_value, int64_t md_lifetime_ms,
+                const char *md_principal_name,
+                const char **extensions, size_t extension_size) {
+        /* Check args for correct format */
+        if (!token_value) {
+		rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG,
+			    "Must supply non-null token value");
+                return -1;
+        }
+        if (!md_principal_name) {
+		rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG,
+			    "Must supply non-null token principal name metadata");
+                return -1;
+        }
+        char errstr[512];
+        if (check_oauthbearer_extension_value(token_value, errstr, sizeof(errstr)) == -1) {
+                rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__INVALID_ARG, "%s", errstr);
+                return -1;
+        }
+        size_t i;
+        for (i = 0; i + 1 < extension_size; i += 2) {
+                if (check_oauthbearer_extension_key(extensions[i], errstr, sizeof(errstr)) == -1 ||
+                        check_oauthbearer_extension_value(extensions[i + 1], errstr, sizeof(errstr)) == -1) {
+                                rd_kafka_set_fatal_error(rk,
+                                        RD_KAFKA_RESP_ERR__INVALID_ARG, "%s", errstr);
+                                return -1;
+                }
+        }
+        rd_ts_t now_ms = rd_clock() / 1000;
```

Comment:
rd_clock() is a monotonic clock, it is not wallclock time.
We can't rely on the application to use the same clock as we do, so either use rd_uclock() (which is the system wall clock time, i.e., gettimeofday), or have the app supply a relative future timeout.
If the timeout is defined by an external system we need to use the wall clock time (to allow for clock adjustments), but if it is a relative time to begin with it should be relative here too.

For absolute time, we should use time_t (sec precision), or struct timeval (usec), as input type to this function, and document that the current timezone of the process will be used to enforce the timeout.

Reply:
👍 Now use `rd_uclock()`.  The definition of the `iat` and `exp` claims is the number of (possibly fractional) seconds since the epoch, so wallclock is the correct one to use.